Add Switch transformers

younesbelkada opened the pull request:
# What does this PR do?

This PR attempts to add Switch Transformers from t5x with @ArthurZucker & @thomwolf 

The architecture seems to be similar to a t5 architecture (the architecture is copied from T5), where the FF layer is slightly modified, introducing the first Mixture of Experts (MoE) architecture inside `transformers` library.

paper: https://arxiv.org/abs/2101.03961
weights: https://github.com/google-research/t5x/blob/eb42c2524bf65c8a46624f1a9b9e034d9bc65b14/docs/models.md#converted-mesh-tensorflow-checkpoints
original modeling code: https://github.com/google/flaxformer/tree/b725bd2a51d70e866d819c92de166fbf24425e6a/flaxformer/architectures/moe


# TODOs:

- [x] Make the forward pass run
- [x] Convert the weights in Pytorch format and upload them on the Hub
- [x] Match the logits between the original implementation and ours

younesbelkada committed:
first commit

younesbelkada committed:
add more comments

younesbelkada committed:
add router v1

younesbelkada committed:
clean up

- remove `tf` modeling files

younesbelkada committed:
clean up

- remove `tf` modeling files

younesbelkada committed:
clean up

younesbelkada committed:
v0 routers

younesbelkada committed:
added more router

- Implemented `ExpertsChooseMaskedRouter`

- added tests
- 2 more routers to implement

younesbelkada committed:
last router

younesbelkada committed:
improved docstring

- completed the docstring in `router.py`
- added more args in the config

younesbelkada committed:
v0 sparse mlp

ArthurZucker committed:
Merge branch 'main' of https://github.com/huggingface/transformers into add_switch_transformers

ArthurZucker committed:
replace wrong naming

younesbelkada committed:
forward pass run

ArthurZucker committed:
update MOE layer

ArthurZucker committed:
small router update

ArthurZucker committed:
fixup

ArthurZucker committed:
consistency

ArthurZucker committed:
remove scatter router

ArthurZucker committed:
remove abstract layer

ArthurZucker committed:
update test and model for integration testing

younesbelkada committed:
v1 conversion

ArthurZucker committed:
update

ArthurZucker committed:
hardcode hack

Commit:
all keys match

ArthurZucker committed:
add gin conversion, without additional libraries

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

ArthurZucker committed:
update conversion sctipy

ArthurZucker committed:
delete router file

ArthurZucker committed:
update tests wrt router deletion

ArthurZucker committed:
fix router issues

ArthurZucker committed:
update expert code

ArthurZucker committed:
update, logits match, code needsREFACTORING

ArthurZucker committed:
Refactor code

Co-authored-by: Younes Belkada <younesbelkada@users.noreply.github.com>

ArthurZucker committed:
add generate tests

Co-authored-by: younesbelkada <younesbelkada@gmail.com>

ArthurZucker committed:
add support for router loss

Co-authored-by: Younes Belkada <younesbelkada@users.noreply.github.com>

younesbelkada committed:
fix forward error

younesbelkada committed:
refactor a bit

younesbelkada committed:
remove `FlaxSwitchTransformers` modules

younesbelkada committed:
more tests pass

ArthurZucker committed:
Update code

Co-authored-by: Younes Belkada <younesbelkada@users.noreply.github.com>

ArthurZucker committed:
fixup

younesbelkada committed:
fix tests

younesbelkada committed:
fix doc

younesbelkada committed:
fix doc + tokenization

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19323). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
fix tokenizer test

younesbelkada committed:
fix test

younesbelkada committed:
fix loss output

ArthurZucker committed:
update code for backward pass

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

younesbelkada committed:
add loss support

ArthurZucker committed:
update documentation

ArthurZucker committed:
fix documentation, clean tokenizer

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

ArthurZucker committed:
more doc fix, cleanup example_switch

younesbelkada committed:
fix failing test

younesbelkada committed:
fix test

younesbelkada committed:
fix test

younesbelkada committed:
fix loss issue

ArthurZucker committed:
move layer

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

ArthurZucker committed:
update doc and fix router capacity usage

ArthurZucker committed:
fixup

ArthurZucker committed:
add sparse mlp index for documentation on hub

ArthurZucker committed:
fixup

ArthurZucker commented on docs/source/en/model_doc/switch_transformers.mdx:
```diff
@@ -0,0 +1,64 @@
+The SwitchTransformers model was proposed in [Switch Transformers: Scaling to Trillion Parameter Models with Simple and Efficient Sparsity](https://arxiv.org/abs/2101.03961) by William Fedus, Barret Zoph, Noam Shazeer.
+
+The Switch Transformer model uses a sparse T5 encoder-decoder architure, where the MLP are replace by a Mixture of Expert (MOE). A routing mecanism (top 1 in this case) associates each token to one of the expert, where each expert is a dense MLP. While switch tranformers have a lot more weights than their equivalent dense models, the sparsity allows for better scaling.
+During a forward pass, only a fraction of the weights are used. The routing mecanism allows the model to select relavant weights on the fly which increases the model capacity. #TODO add the intuition about moving the loss curve.
```
```suggestion
During a forward pass, only a fraction of the weights are used. The routing mecanism allows the model to select relevant weights on the fly which increases the model capacity without increasing the number of operations.
```

ArthurZucker commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,233 @@
+
+logger = logging.get_logger(__name__)
+
+SWITCH_TRANSFORMERS_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "HFLAY/switch_base_8": "https://huggingface.co/HFLAY/switch_base_8/blob/main/config.json",
+}
```
```suggestion
SWITCH_TRANSFORMERS_PRETRAINED_CONFIG_ARCHIVE_MAP = {
     "google/switch-base-8": "https://huggingface.co/google/switch-base-8/blob/main/config.json",
}
```

ArthurZucker commented on src/transformers/models/switch_transformers/convert_switch_transformers_original_flax_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,231 @@
+logging.set_verbosity_info()
+
+
+MODEL_MAPPING = {
+    "switch_base_8": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_16": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_32": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_64": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_128": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_256": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_512": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_1024": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+    "switch_base_2048": [
+        "https://github.com/google/flaxformer/tree/main/flaxformer/t5x/configs/moe/models/switch_base.gin"
+    ],
+}
```
```suggestion
```

ArthurZucker committed:
test sparse mix architecture

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1894 @@
+            hidden_states (`torch.Tensor`) :
+                [num_groups, tokens_per_group, hidden_dim] inputs to send to experts.
+        Returns:
+            Router indices or mask torch.Tensors (depending on router type).
```
```suggestion
            Tuple[`torch.Tensor`, `torch.Tensor`, `torch.Tensor`] Tuple containing the expert index, the router probs and the router logits. The router probabilities and logits are required to compute the loss. 
```

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1894 @@
+        """
+        router_probs, router_logits = self._compute_router_probabilities(hidden_states)
+
+        # Flax code for reference TODO check what happens with padded inputs here
```
Has to be adressed

sgugger replied:
Flagging this just in case!

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1894 @@
+            )
+
+    def forward(self, hidden_states):
+        r"""
+        Hold on, this will be slightly tricky to understand In the correct order, a MoE layer does the following:
+
+        1- Gets the `router_mask` from the router. This mask will contain the indices of the routed tokens. Also
+        retrieve the probabilities (max prob) for each token. The probabilities are needed in the computation of the
+        hidden states since the probabilities will be broadcasted to the hidden states values (they can be interpreted
+        as a scaling factor).
+
+        2- Dispatch the tokens to the experts.
+
+        """
```
Update the args here

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1894 @@
+
+        outputs = outputs + (router_tuple,)
+
+        return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
```
```suggestion
        return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights), (router_tuple)
```

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1894 @@
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
```
```suggestion
        output_hidden_states (`bool`, *optional*):
            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
            more detail.
        output_router_logits (`bool`, *optional*):
            Whether or not to return the logits of all the routers. They are useful for computing the router loss, and
            should not be returned during inference.
```

ArthurZucker reviewed (commented):
Still a bit more work to do on documentation but overall I think we made a lot of progress today! Especially with the router expert capacity issue 🥳

ArthurZucker committed:
Apply suggestions from code review

ArthurZucker committed:
Update docs/source/en/model_doc/switch_transformers.mdx

ArthurZucker committed:
fixup on update

younesbelkada committed:
fix tests

younesbelkada committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

younesbelkada committed:
fix another test

younesbelkada committed:
attempt fix

younesbelkada committed:
Update src/transformers/models/switch_transformers/configuration_switch_transformers.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/models/switch_transformers/convert_switch_transformers_original_flax_checkpoint_to_pytorch.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
try

younesbelkada committed:
all tests pass

younesbelkada committed:
fix jitter noise

younesbelkada commented on src/transformers/modeling_outputs.py:
```diff
@@ -347,6 +432,77 @@ class Seq2SeqModelOutput(ModelOutput):
+        encoder_router_logits (`tuple(torch.FloatTensor)`, *optional*, returned when `output_router_logits=True` is passed or when `config.add_router_probs=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, sequence_length, num_experts)`.
+
+            Router logits of the encoder model, useful to compute the auxiliary loss for Mixture of Experts models.
```
```suggestion
            Router logits of the encoder model, useful to compute the auxiliary loss and the z_loss for the sparse modules.
```

younesbelkada commented on src/transformers/modeling_outputs.py:
```diff
@@ -581,6 +737,80 @@ class Seq2SeqLMOutput(ModelOutput):
+        encoder_router_logits (`tuple(torch.FloatTensor)`, *optional*, returned when `output_router_logits=True` is passed or when `config.add_router_probs=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, sequence_length, num_experts)`.
+
+            Router logits of the encoder model, useful to compute the auxiliary loss for Mixture of Experts models.
```
```suggestion
            Router logits of the encoder model, useful to compute the auxiliary loss and z_loss for Mixture of Experts models.
```

younesbelkada commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,233 @@
+    Arguments:
+        vocab_size (`int`, *optional*, defaults to 32128):
+            Vocabulary size of the SwitchTransformers model. Defines the number of different tokens that can be
+            represented by the `inputs_ids` passed when calling [`SwitchTransformersModel`] or
+            [`FlaxSwitchTransformersModel`].
```
```suggestion
            [`FlaxSwitchTransformersModel`].
```
```suggestion
            represented by the `inputs_ids` passed when calling [`SwitchTransformersModel`].
```

younesbelkada commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,233 @@
+            Number of attention heads for each attention layer in the Transformer encoder.
+        num_experts (`int`, *optional*, defaults to 8):
+            Number of experts for each SwitchTransformer layer.
+        router_type (`str`, *optional*, defaults to `tokens_masked`):
+            Router type - choice between `tokens_masked` and `tokens_scatter`, `experts_masked`.
```
```suggestion
        router_type (`str`, *optional*, defaults to `tokens_masked`):
            Router type - choice between `tokens_masked` and `tokens_scatter`, `experts_masked`.
```
Do we need this now since we're using a single router type?

ArthurZucker replied:
lets keep it

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1909 @@
+        for idx in range(config.num_experts):
+            self.experts[f"expert_{idx}"] = expert_class(config)
+
+    def _get_router(self, config):
```
We should probably discuss if this function is needed after all

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1909 @@
+        hidden states since the probabilities will be broadcasted to the hidden states values (they can be interpreted
+        as a scaling factor).
+
+        2- Dispatch the tokens to the experts.
```
```suggestion
        2- Dispatch the tokens to the experts. We do a classic for loop over the experts and assign for each expert
        the corresponding hidden state
```

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1909 @@
+
+SWITCH_TRANSFORMERS_START_DOCSTRING = r"""
+
+    The SWITCH_TRANSFORMERS model was proposed in [Exploring the Limits of Transfer Learning with a Unified
+    Text-to-Text Transformer](https://arxiv.org/abs/1910.10683) by Colin Raffel, Noam Shazeer, Adam Roberts, Katherine
+    Lee, Sharan Narang, Michael Matena, Yanqi Zhou, Wei Li, Peter J. Liu. It's an encoder decoder transformer
+    pre-trained in a text-to-text denoising generative setting.
```
```suggestion
    The SWITCH_TRANSFORMERS model was proposed in [Switch Transformers: Scaling to Trillion Parameter Models with Simple and Efficient Sparsity](https://arxiv.org/abs/2101.03961) by [William Fedus](https://arxiv.org/search/cs?searchtype=author&query=Fedus%2C+W), [Barret Zoph](https://arxiv.org/search/cs?searchtype=author&query=Zoph%2C+B), and [Noam Shazeer](https://arxiv.org/search/cs?searchtype=author&query=Shazeer%2C+N). It's an encoder-decoder T5-like model with sparse Feed Forward that stands for Mixture of Experts (MoE) architecture.
```

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
````diff
@@ -0,0 +1,1909 @@
+        ```python
+        >>> from transformers import T5Tokenizer, SwitchTransformersModel
+
+        >>> tokenizer = T5Tokenizer.from_pretrained("ybelkada/switch_transformers-base")
+        >>> model = SwitchTransformersModel.from_pretrained("ybelkada/switch_transformers-base")
````
```suggestion
        >>> tokenizer = T5Tokenizer.from_pretrained("google/switch-base-8")
        >>> model = SwitchTransformersModel.from_pretrained("google/switch-base-8")
```

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
````diff
@@ -0,0 +1,1909 @@
+        ```python
+        >>> from transformers import T5Tokenizer, SwitchTransformersForConditionalGeneration
+
+        >>> tokenizer = T5Tokenizer.from_pretrained("ybelkada/switch_transformers-base")
+        >>> model = SwitchTransformersForConditionalGeneration.from_pretrained("ybelkada/switch_transformers-base")
````
```suggestion
        >>> tokenizer = T5Tokenizer.from_pretrained("google/switch-base-8")
        >>> model = SwitchTransformersForConditionalGeneration.from_pretrained("google/switch-base-8")
```

younesbelkada commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
````diff
@@ -0,0 +1,1909 @@
+        ```python
+        >>> from transformers import T5Tokenizer, SwitchTransformersEncoderModel
+
+        >>> tokenizer = T5Tokenizer.from_pretrained("ybelkada/switch_transformers-base")
+        >>> model = SwitchTransformersEncoderModel.from_pretrained("ybelkada/switch_transformers-base")
````
```suggestion
        >>> tokenizer = T5Tokenizer.from_pretrained("google/switch-base-8")
        >>> model = SwitchTransformersEncoderModel.from_pretrained("google/switch-base-8")
```

younesbelkada reviewed (commented):
Added few improvements of the documentation!

younesbelkada committed:
Apply suggestions from code review

younesbelkada committed:
doc tests pass

ArthurZucker commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+
+    def __init__(
+        self,
+        vocab_size=32128,
+        d_model=768,
+        d_kv=64,
+        d_ff=2048,
+        num_layers=12,
+        num_sparse_encoder_layers=3,
+        num_decoder_layers=12,
+        num_sparse_decoder_layers=3,
+        num_heads=12,
+        num_experts=8,
+        expert_capacity=64,
+        router_type="tokens_masked",
+        router_bias=False,
+        router_jitter_noise=0.01,
+        router_dtype="float32",
+        router_ignore_padding_tokens=False,
+        relative_attention_num_buckets=32,
+        relative_attention_max_distance=128,
+        dropout_rate=0.1,
+        router_z_loss_coef=0.001,
+        router_aux_loss_coef=0.001,
+        layer_norm_epsilon=1e-6,
+        initializer_factor=1.0,
+        feed_forward_proj="relu",
+        is_encoder_decoder=True,
+        add_router_probs=False,
+        use_cache=True,
+        pad_token_id=0,
+        eos_token_id=1,
```
The order does not correspond with the doc

younesbelkada replied:
FIxed in 16e23c4bd4456ffbdf122f9fce160086514e5864

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1911 @@
+        next_states = hidden_states.clone()
+        for idx, expert in enumerate(self.experts.values()):
+
+            # 1. Get the index of the tokens that are routed to the current expert
+            # masked_indices has a shape of `batch_size`, `seq_len`, `num_experts`
+            token_indices = router_mask[:, :, idx].bool()
+
+            # 2. Update only the hidden states affected by the routing
+            next_states[token_indices] = expert(hidden_states[token_indices])
```
```suggestion
            token_indices = router_mask[:, :, idx].bool()
            next_states[token_indices] = expert(hidden_states[token_indices])
```

ArthurZucker commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1911 @@
+                    layer_past_state.index_select(0, beam_idx.to(layer_past_state.device)),
+                )
+
+            assert reordered_layer_past_states[0].shape == layer_past_states[0].shape
+            assert len(reordered_layer_past_states) == len(layer_past_states)
```
Let's  raise an error

younesbelkada replied:
Fixed in 88e68b5ef8b35f693cb25a042276e1390e1eb1a4

younesbelkada committed:
Update src/transformers/models/switch_transformers/modeling_switch_transformers.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/models/switch_transformers/modeling_switch_transformers.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
remove assert

younesbelkada committed:
change config order

younesbelkada committed:
Merge remote-tracking branch 'upstream_2/main' into add_switch_transformers

younesbelkada committed:
fix readme japanese

sgugger commented on docs/source/en/model_doc/switch_transformers.mdx:
```diff
@@ -0,0 +1,64 @@
+
+Tips:
+
+- SwitchTransformers uses the T5Tokenizer, which can be loaded directly from each model's repository.
```
```suggestion
- SwitchTransformers uses the [`T5Tokenizer`], which can be loaded directly from each model's repository.
```

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+            Number of attention heads for each attention layer in the Transformer encoder.
+        num_experts (`int`, *optional*, defaults to 8):
+            Number of experts for each SwitchTransformer layer.
+        router_type (`str`, *optional*, defaults to `tokens_masked`):
```
```suggestion
        router_type (`str`, *optional*, defaults to `"tokens_masked"`):
```

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+        num_experts (`int`, *optional*, defaults to 8):
+            Number of experts for each SwitchTransformer layer.
+        router_type (`str`, *optional*, defaults to `tokens_masked`):
+            Router type - choice between `tokens_masked` and `tokens_scatter`, `experts_masked`.
```
```suggestion
            Router type - choose between `"tokens_masked", `"tokens_scatter"` and `"experts_masked"`.
```

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+        router_jitter_noise (`float`, *optional*, defaults to 0.1):
+            Amount of noise to add to the router.
+        router_dtype (`str`, *optional*, default to `float32`):
+            The `dtype` used for the routers. It is preferable to keep the `dtype` to `float32` as specified in the
```
```suggestion
            The `dtype` used for the routers. It is preferable to keep the `dtype` to `"float32"` as specified in the
```

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+            Amount of noise to add to the router.
+        router_dtype (`str`, *optional*, default to `float32`):
+            The `dtype` used for the routers. It is preferable to keep the `dtype` to `float32` as specified in the
+            "selective precision" discussion in https://arxiv.org/abs/2101.03961.
```
```suggestion
            *selective precision* discussion in [the paper](https://arxiv.org/abs/2101.03961).
```

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,232 @@
+        )
+
+
+class SwitchTransformersOnnxConfig(OnnxSeq2SeqConfigWithPast):
+    @property
+    def inputs(self) -> Mapping[str, Mapping[int, str]]:
+        common_inputs = {
+            "input_ids": {0: "batch", 1: "encoder_sequence"},
+            "attention_mask": {0: "batch", 1: "encoder_sequence"},
+        }
+        if self.use_past:
+            common_inputs["attention_mask"][1] = "past_encoder_sequence + sequence"
+            common_inputs["decoder_input_ids"] = {0: "batch"}
+            common_inputs["decoder_attention_mask"] = {0: "batch", 1: "past_decoder_sequence + sequence"}
+        else:
+            common_inputs["decoder_input_ids"] = {0: "batch", 1: "decoder_sequence"}
+            common_inputs["decoder_attention_mask"] = {0: "batch", 1: "decoder_sequence"}
+
+        if self.use_past:
+            self.fill_with_past_key_values_(common_inputs, direction="inputs")
+
+        return common_inputs
+
+    @property
+    def default_onnx_opset(self) -> int:
+        return 13
```
Usually the Onnx config is added in a second step. Let's maybe keep this for another PR?

younesbelkada replied:
Fixed in 09788717502af1cbd8dea7562c16e63d82db6644

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -509,6 +512,7 @@
         ("pegasus_x", "PegasusXForConditionalGeneration"),
         ("plbart", "PLBartForConditionalGeneration"),
         ("prophetnet", "ProphetNetForConditionalGeneration"),
+        ("switch_transformers", "SwitchTransformersForConditionalGeneration"),
```
Please also add the default to `T5Tokenizer` in the tokenization_auto module.

younesbelkada replied:
Fixed in ccc28a9f1652aaedf9590842b1d624cc697acd58

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1914 @@
+        self.classifier = nn.Linear(config.hidden_size, self.num_experts, bias=config.router_bias)
+        self.jitter_noise = config.router_jitter_noise
+        self.ignore_padding_tokens = config.router_ignore_padding_tokens
+        self.dtype = getattr(torch, config.router_dtype)
```
Nit: might be good to add some defense mechanism in the config for typos, so that the user gets a clear error message.

younesbelkada replied:
Agreed! Proposed a solution in a2d786b4b3f24409bb596ee1868d2ec9cb3698f6

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1914 @@
+            - Scatter Routers
+
+        """
+        if config.router_type.lower() == "tokens_masked":
+            return SwitchTransformersTop1Router(config)
+        else:
+            raise NotImplementedError(
+                f"{config.router_type.lower()} not implemented ! Please chose a router in "
+                "`{'tokens_masked','experts_masked'}`"
+            )
```
If only one router_type is supported in the paper for Switch Transformers, let's remove the attribute from the config and this logic to just use the router they support. Transformers is not a modular toolbox :-)

younesbelkada replied:
Agreed! Fixed in ef1fa19beb6687ffb913bf560b8eb7b7508936ab

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1914 @@
+        decoder_start_token_id = self.config.decoder_start_token_id
+        pad_token_id = self.config.pad_token_id
+
+        assert decoder_start_token_id is not None, (
+            "self.model.config.decoder_start_token_id has to be defined. In SwitchTransformers it is usually set to"
+            " the pad_token_id. See SwitchTransformers docs for more information"
+        )
```
No new assert in the code base please ;-). Use a test and raise an appropriate error.

younesbelkada replied:
Thanks! Fixed in e2dc2b6a417335fdf14c9b4ed5322a415ba4aa74

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1914 @@
+        r"encoder.embed_tokens.weight",
+        r"decoder.embed_tokens.weight",
+    ]
+    _keys_to_ignore_on_load_unexpected = [
+        r"decoder.block.0.layer.1.EncDecAttention.relative_attention_bias.weight",
+    ]
```
Why would the model have this key? Can we just remove it from the checkpoint?

younesbelkada replied:
I think we kept it because `T5` relies on that too, and the model has this key since [only the first block of the decoder](https://github.com/younesbelkada/transformers/blob/ab67f48078be72c78200604c2023f4590c32e3e8/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L940) layer has this module.
I can confirm the CI test when removing that from `_keys_to_ignore_on_load_unexpected`

sgugger replied:
It's to not raise a warning if the weight is in the checkpoint but not loaded in the model. I don't see a config key to ignore setting this bias in the first block of the encoder, so it will always be there, no? Do you see any warning popping when removing the key and loading a pretrained model?

younesbelkada replied:
Ah I see, yes that makes sense. Yes this weight will be always there, I think that we can safely remove it then

younesbelkada replied:
Removed it in 437eef7acbce56c1d70774c22474775ceda075a8

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+        (SwitchTransformersModel, SwitchTransformersForConditionalGeneration) if is_torch_available() else ()
+    )
+    all_generative_model_classes = (SwitchTransformersForConditionalGeneration,) if is_torch_available() else ()
+    all_parallelizable_model_classes = (
```
Should be removed as it corresponds to the old `parallelize` API that I don't see implemented here (and it's normal since it's deprecated).

younesbelkada replied:
Fixed in 0240906ec6148c9996d535d127626819e8e2e2ca

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+            is_encoder_decoder=self.is_encoder_decoder,
+        )
+
+        return (
+            config,
+            input_ids,
+            attention_mask,
+        )
```
Fits in one line.

younesbelkada replied:
Fixed in 0240906ec6148c9996d535d127626819e8e2e2ca

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+            attention_mask,
+        )
+
+    def create_and_check_model(
+        self,
+        config,
+        input_ids,
+        attention_mask,
+    ):
```
Fits in one line.

younesbelkada replied:
Fixed in 0240906ec6148c9996d535d127626819e8e2e2ca !

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+
+        self.parent.assertEqual(encoder_output.size(), (self.batch_size, self.encoder_seq_length, self.hidden_size))
+
+    def create_and_check_model_fp16_forward(
+        self,
+        config,
+        input_ids,
+        attention_mask,
+    ):
```
Here too probably

younesbelkada replied:
Fixed in 0240906ec6148c9996d535d127626819e8e2e2ca

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            input_ids,
+            attention_mask,
+        ) = config_and_inputs
```
This fits in one line too.

younesbelkada replied:
Fixed in 0240906ec6148c9996d535d127626819e8e2e2ca

sgugger commented on tests/models/switch_transformers/test_modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1197 @@
+    test_pruning = False
+    test_resize_embeddings = False
+    test_model_parallel = True
+    all_parallelizable_model_classes = (SwitchTransformersEncoderModel,) if is_torch_available() else ()
```
As above, to remove.

younesbelkada replied:
Should be addressed in 0240906ec6148c9996d535d127626819e8e2e2ca

sgugger reviewed (commented):
Very clean, thanks a lot for adding this new model!
My only main comment is around the router type in the config/then `_get_router` funciton that does not seem to be used since only one type is supported. It should be removed and the code simplified to only support said router!

Also don't forget to use a spellchecker on the doc files ;-)

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
remove parallelizable tests + add one liners

younesbelkada committed:
remove ONNX config

younesbelkada committed:
fix nits

- add `T5Tokenizer` in auto mapping
- remove `Switch Transformers` from ONNX supported models

younesbelkada committed:
remove `_get_router`

younesbelkada committed:
remove asserts

younesbelkada committed:
add check in test for `router_dtype`

younesbelkada committed:
add `SwitchTransformersConfig` in `run_pipeline_test`

sgugger commented on tests/pipelines/test_pipelines_summarization.py:
```diff
@@ -54,7 +55,7 @@ def run_pipeline_test(self, summarizer, _):
         self.assertEqual(outputs, [{"summary_text": ANY(str)}])
 
-        if not isinstance(model.config, (T5Config, LongT5Config, LEDConfig)):
+        if not isinstance(model.config, (SwitchTransformersConfig, T5Config, LongT5Config, LEDConfig)):
```
Modify the comment below a well :-)

younesbelkada replied:
Great! Fixed in 7c3e5aabf6cd72b84e24af34d377fc1310cdca8f

younesbelkada commented on tests/pipelines/test_pipelines_summarization.py:
```diff
@@ -54,7 +55,7 @@ def run_pipeline_test(self, summarizer, _):
 
-        if not isinstance(model.config, (T5Config, LongT5Config, LEDConfig)):
+        if not isinstance(model.config, (SwitchTransformersConfig, T5Config, LongT5Config, LEDConfig)):
             # LED, T5, LongT5 can handle it.
```
```suggestion
            # Switch Transformers, LED, T5, LongT5 can handle it.
```

younesbelkada committed:
Update tests/pipelines/test_pipelines_summarization.py

ArthurZucker committed:
add huge model conversion script

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

younesbelkada committed:
fix slow tests

- add better casting for `Linear8bitLt`
- remove `torchscript` tests

ArthurZucker committed:
add make dir

ArthurZucker committed:
Merge branch 'add_switch_transformers' of https://github.com/younesbelkada/transformers into add_switch_transformers

ArthurZucker committed:
style on new script

younesbelkada commented:
Thanks a lot @sgugger for your comments! 
Would love to have another round of review as I added some modification for `accelerate` and `bnb` compatibility 🙏

younesbelkada committed:
fix nits

- doctest
- remove `_keys_to_ignore_on_load_unexpected`

sgugger commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,206 @@
+            raise ValueError(f"`router_dtype` must be one of 'float32', 'float16' or 'bfloat16', got {router_dtype}")
+        self.router_dtype = router_dtype
+
+        if router_dtype not in ["float16", "float32", "bfloat16"]:
+            raise ValueError("""Please select a correct `router_dtype` from ["float16", "float32", "bfloat16"].""")
```
Seems like this one can be removed now.

younesbelkada replied:
Fixed in f82f0cf7bb434b5c113bf70f7264875cc289b18e !

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1907 @@
+        mask_seq_length = past_key_values[0][0].shape[2] + seq_length if past_key_values is not None else seq_length
+
+        if use_cache is True:
+            assert self.is_decoder, f"`use_cache` can only be set to `True` if {self} is used as a decoder"
```
One assert to switch to a test + exception (since this is not a copied from :-) )

younesbelkada replied:
Removed all assert statements in 2dd9cff !

sgugger commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1907 @@
+    """SWITCH_TRANSFORMERS Model with a `language modeling` head on top.""", SWITCH_TRANSFORMERS_START_DOCSTRING
+)
+class SwitchTransformersForConditionalGeneration(SwitchTransformersPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [
+        r"encoder.embed_tokens.weight",
+        r"decoder.embed_tokens.weight",
+        r"lm_head.weight",
+    ]
```
Nit: can take less vertical space.

younesbelkada replied:
Managed to change it for the attributes above in [16e7ff5](https://github.com/huggingface/transformers/pull/19323/commits/16e7ff53971730ba39608c59881697f7c68e4afa) but not for this one (not sure why the decorator above is not affected by the vertical formatting though 🤔

sgugger reviewed (approved):
Looking great!

LysandreJik commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,206 @@
+# coding=utf-8
+# Copyright 2020, The SwitchTransformers Authors and HuggingFace Inc.
```
Would that be Google given the original codebase google/flaxformers?

younesbelkada replied:
.. and in 2f5806938d95a2ff819e240e18f0b50dd456edcb

LysandreJik commented on src/transformers/models/switch_transformers/modeling_switch_transformers.py:
```diff
@@ -0,0 +1,1907 @@
+        }
+        return dummy_inputs
+
+    def _init_weights(self, module):
```
Impressive method!

LysandreJik reviewed (approved):
Impressive contribution! Great to have so many pretrained checkpoints at release

younesbelkada commented on src/transformers/models/switch_transformers/configuration_switch_transformers.py:
```diff
@@ -0,0 +1,206 @@
+            raise ValueError(f"`router_dtype` must be one of 'float32', 'float16' or 'bfloat16', got {router_dtype}")
+        self.router_dtype = router_dtype
+
+        if router_dtype not in ["float16", "float32", "bfloat16"]:
+            raise ValueError("""Please select a correct `router_dtype` from ["float16", "float32", "bfloat16"].""")
```
```suggestion
```

younesbelkada committed:
Update src/transformers/models/switch_transformers/configuration_switch_transformers.py

younesbelkada committed:
add google as authors

younesbelkada committed:
fix year

younesbelkada committed:
remove last `assert` statements

younesbelkada committed:
standardize vertical spaces

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_switch_transformers

younesbelkada committed:
fix failing import

younesbelkada committed:
fix another failing test

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19323). All of your documentation changes will be reflected on that endpoint.

ArthurZucker committed:
Remove strange àuthorized_keys`

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19323). All of your documentation changes will be reflected on that endpoint.

ArthurZucker committed:
removing todo and padding that is never used

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19323). All of your documentation changes will be reflected on that endpoint.

younesbelkada commented:
Failing tests seems to be unrelated to this PR, merging!
